src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala

```diff
@@ -132,16 +150,217 @@ object CodeGenYul extends CodeGenerator {
+        val f: (VariableDeclWithSpec) => TypedName = v => TypedName(v.varName, mapObsTypeToABI(v.typIn.toString()))
+        val parameters: Seq[TypedName] = transaction.args.map[TypedName](f)
+
+        val ret: Seq[TypedName] =
+            if (transaction.retType.isEmpty) {
+                Seq()
+            } else { // TODO hard code return variable name now
+                Seq(TypedName("retValTempName", mapObsTypeToABI(transaction.retType.get.toString())) )
+            }
+
+        var body: Seq[YulStatement] = Seq()
+        for (s <- transaction.body){
+            body = body ++ translateStatement(s)
+        }
+
+        Seq(FunctionDefinition(
+            transaction.name, // TODO rename transaction name (by adding prefix/suffix)
+            parameters,
+            ret,
+            Block(body)))
+    }
+
+    def translateStatement(s: Statement): Seq[YulStatement] = {
+        s match {
+            case Return() =>
+                Seq()
+            case ReturnExpr(e) =>
+                translateExpr(e)
+            case Assignment(assignTo, e) =>
+                assignTo match {
+                    // TODO only support int/int256 now
+                    case ReferenceIdentifier(x) =>
+                        val idx = temp_symbol_table(x)
+                        val value = e match {
+                            case NumLiteral(v) => v
+                            case _ =>
+                                assert(false, "TODO")
+                                0
+                        }
+                        val expr = FunctionCall(Identifier("sstore"), Seq(Literal(LiteralKind.number, idx.toString(), "int"),Literal(LiteralKind.number, value.toString(), "int")))
+                        Seq(ExpressionStatement(expr))
+                    case _ =>
+                        assert(false, "TODO")
+                        Seq()
+                }
+
+            case _ =>
+                assert(false, "TODO")
+                Seq()
+
+        }
+    }
+
+    def translateExpr(e: Expression): Seq[YulStatement] = {
+        e match {
+            case ReferenceIdentifier(x) =>
+                val idx = temp_symbol_table(x)
+                val expr = FunctionCall(Identifier("sload"), Seq(Literal(LiteralKind.number, idx.toString(), "int")))
+                Seq(ExpressionStatement(expr))
+            case _ => Seq() // TODO unimplemented
+        }
     }
 }
 
 // temporary function, not designed for a full recursive walk through of the object
-class Scope(val mainContractName: String) {
-    val creationObject:String = mainContractName
-    val runtimeObject:String = mainContractName + "_deployed"
-}
\ No newline at end of file
+class ObjScope(obj: YulObject) {
+    class Func(val code: String){}
+    class Case(val hash: String){}
+    class Call(val call: String){}
+
+    // TODO unimplemented; hardcode to uint256 for now
+    def mapObsTypeToABI(ntype: String): String = {
+        "uint256"
```

Comment:
Why is the input of String type?

Reply:
so far we use String to record what the type is.